### PR TITLE
Dev doc enhancements

### PR DESCRIPTION
**Pull request by ian-r-rose**
## References

Alleviates #6463 

## Code changes

None
## User-facing changes

Better extension dev docs

## Backwards-incompatible changes

None

### CODE REVIEW

File: docs/source/developer/extension_dev.rst [hunk 1]
```diff
@@ -309,17 +304,34 @@ with each other through typed interfaces that are provided by ``Token`` objects.
...
 
+- ``@jupyterlab/application:IConnectionLost``: A token for invoking the dialog shown
+  when JupyterLab has lost its connection to the server. Use this if, for some reason,
+  you want to bring up the "connection lost" dialog under new circumstances.
+- ``@jupyterlab/application:IInfo``: A token providing metadata about the current
+  application, including currently disabled extensions and whether dev mode is enabled.
+- ``@jupyterlab/application:IPaths``: A token providing information about various
+  URLs and server paths for the current application. Use this token if you want to
+  assemble URLs to use the JupyterLab REST API.
+- ``@jupyterlab/application:ILabStatus``: An interface for interacting with tge application busy/dirty
```
Comment (vidartf) [inline docs/source/developer/extension_dev.rst]:
```suggestion
- ``@jupyterlab/application:ILabStatus``: An interface for interacting with the application busy/dirty
```

**Comment by ian-r-rose** [comment]
I should note that I don't think this needs to be in for the RC, as it's documentation only.

**Comment by ian-r-rose** [comment]
Okay, some feedback would be useful at this point. This does the following:
1. A bunch of rewording, formatting, and general cleanup.
2. Removing some (extremely) out-of-date documentation
3. Adds one big new page discussing and showing examples of common extension points for extension authors (commands, menus, etc).

**Comment by jtpio** [comment]
Regarding the Mime Renderer extensions, it looks like the JavaScript cookiecutter is now deprecated (archived): https://github.com/jupyterlab/mimerender-cookiecutter

Something that can be removed from the docs then?

**Comment by ian-r-rose** [comment]
Good point @jtpio, I've removed the reference.

File: docs/source/developer/extension_dev.rst [hunk 2]
```diff
@@ -241,9 +239,8 @@ Mime Renderer Extensions
 
 Mime Renderer extensions are a convenience for creating an extension
 that can render mime data and potentially render files of a given type.
-We provide cookiecutters for Mime render extensions in
-`JavaScript <https://github.com/jupyterlab/mimerender-cookiecutter>`__ and
-`TypeScript <https://github.com/jupyterlab/mimerender-cookiecutter-ts>`__.
+We provide a cookiecutters for mime renderer extensions in TypeScript
```
Comment (jtpio) [inline docs/source/developer/extension_dev.rst]:
```suggestion
We provide a cookiecutter for mime renderer extensions in TypeScript
```

File: docs/source/developer/extension_dev.rst [hunk 3]
```diff
@@ -315,13 +308,30 @@ with each other through typed interfaces that are provided by ``Token`` objects.
...
+  application, including currently disabled extensions and whether dev mode is enabled.
+- ``@jupyterlab/application:IPaths``: A token providing information about various
+  URLs and server paths for the current application. Use this token if you want to
+  assemble URLs to use the JupyterLab REST API.
+- ``@jupyterlab/application:ILabStatus``: An interface for interacting with the application busy/dirty
+  status. Use this if you want to set the application "busy" favicon, or to set
+  the application "dirty" status, which asks the user for confirmation before leaving.
+- ``@jupyterlab/application:ILabShell``: An interface to the JupyterLab shell.
+  The top-level application object also has a reference to the shell, but it has a restricted
+  interface in order to be agnostic to different spins on the applcation.
```
Comment (jtpio) [inline docs/source/developer/extension_dev.rst]: Quick typo:
```suggestion
  interface in order to be agnostic to different spins on the application.
```

File: docs/source/developer/extension_dev.rst [hunk 4]
```diff
@@ -391,17 +409,21 @@ might want to use them in your extensions.
...
   created by the application.
+- ``@jupyterlab/statusbar:IStatusBar``: An interface to the status bar on the application.
+  Use this if you want to add new status bar items.
 - ``@jupyterlab/terminal:ITerminalTracker``: A widget tracker for terminals.
   Use this if you want to be able to iterate over and interact with terminals
   created by the application.
 - ``@jupyterlab/tooltip:ITooltipManager``: An interface to the tooltip manager for the application.
   Use this to allow your extension to invoke a tooltip.
+- ``@jupyterlab/vdom:IVDOMTracker``: A widget tracker for virtual DOM (VDOM) documents.
+  Use the to iterate over and interact with VDOM instances created by the application.
```
Comment (jtpio) [inline docs/source/developer/extension_dev.rst]:
```suggestion
  Use this to iterate over and interact with VDOM instances created by the application.
```

File: docs/source/developer/extension_points.rst [hunk 5]
```diff
@@ -0,0 +1,355 @@
+.. _developer-extension-points:
+
+Common Extension Points
+-----------------------
+
+Most of the component parts of JupyterLab are designed to be extensible,
+and they provide public APIs via that can be requested in extensions via tokens.
```
Comment (jtpio) [inline docs/source/developer/extension_points.rst]:
```suggestion
and they provide public APIs that can be requested in extensions via tokens.
```

File: docs/source/developer/extension_points.rst [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+A list of tokens that extension authors can request is documented in :ref:`tokens`.
+
+This is intended to be a guide for some of JupyterLab's most commonly-used extension points.
+However, it is not an exhaustive account of how to extend the application components,
```
Comment (jtpio) [inline docs/source/developer/extension_points.rst]: Trailing comma?

**Comment by ian-r-rose** [comment]
Thanks for the proofread, @jtpio!

**Comment by jtpio** [comment]
Looks good to me.

Anything else to add or it's ready to be merged?

**Comment by ian-r-rose** [comment]
I'm happy with it for the moment. There's more work to be done in docs, but it seems reasonable to do that in follow-ups.

**Comment by afshin** [comment]
Thanks @ian-r-rose! Thanks for reviewing, @jtpio!

**Pull request merged**